python/Dockerfile

```diff
@@ -32,43 +34,89 @@ RUN apt-get update \
-RUN pyenv install $PY_3_8 \
+RUN mkdir -p /opt/hostedtoolcache/Python/$PY_3_8/x64/ \
+  && cd /opt/hostedtoolcache/Python/$PY_3_8/x64/ \
+  && curl -L https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json \
+  | jq 'map(select(.version == '\"$PY_3_8\"')) [].files | map(select(.platform == "linux" and .arch == "x64" and .platform_version == "22.04"))[] | .download_url' \
+  | xargs -i -- curl -L {} \
+  | tar xz \
+  && rm build_output.txt Python-$PY_3_8.tgz tools_structure.txt setup.sh \
+  && mkdir /usr/local/.pyenv/versions/ \
+  && ln -s /opt/hostedtoolcache/Python/$PY_3_8/x64 /usr/local/.pyenv/versions/$PY_3_8 \
   && bash /opt/python/helpers/build $PY_3_8 \
-  && cd /usr/local/.pyenv \
-  && tar czf 3.8.tar.gz versions/$PY_3_8
+  && cd /opt/hostedtoolcache/Python \
+  && tar czf $PY_3_8.tar.gz $PY_3_8
 
 FROM python-core as python-3.9
-RUN pyenv install $PY_3_9 \
+RUN mkdir -p /opt/hostedtoolcache/Python/$PY_3_9/x64/ \
+  && cd /opt/hostedtoolcache/Python/$PY_3_9/x64/ \
+  && curl -L https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json \
+  | jq 'map(select(.version == '\"$PY_3_9\"')) [].files | map(select(.platform == "linux" and .arch == "x64" and .platform_version == "22.04"))[] | .download_url' \
+  | xargs -i -- curl -L {} \
+  | tar xz \
+  && rm build_output.txt Python-$PY_3_9.tgz tools_structure.txt setup.sh \
+  && mkdir /usr/local/.pyenv/versions/ \
+  && ln -s /opt/hostedtoolcache/Python/$PY_3_9/x64 /usr/local/.pyenv/versions/$PY_3_9 \
   && bash /opt/python/helpers/build $PY_3_9 \
-  && cd /usr/local/.pyenv \
-  && tar czf 3.9.tar.gz versions/$PY_3_9
+  && cd /opt/hostedtoolcache/Python \
+  && tar czf $PY_3_9.tar.gz $PY_3_9
 
 FROM python-core as python-3.10
-RUN pyenv install $PY_3_10 \
+RUN mkdir -p /opt/hostedtoolcache/Python/$PY_3_10/x64/ \
+  && cd /opt/hostedtoolcache/Python/$PY_3_10/x64/ \
+  && curl -L https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json \
+  | jq 'map(select(.version == '\"$PY_3_10\"')) [].files | map(select(.platform == "linux" and .arch == "x64" and .platform_version == "22.04"))[] | .download_url ' \
+  | xargs -i -- curl -L {} \
+  | tar xz \
+  && rm build_output.txt Python-$PY_3_10.tgz tools_structure.txt setup.sh \
+  && mkdir /usr/local/.pyenv/versions/ \
+  && ln -s /opt/hostedtoolcache/Python/$PY_3_10/x64 /usr/local/.pyenv/versions/$PY_3_10 \
   && bash /opt/python/helpers/build $PY_3_10 \
-  && cd /usr/local/.pyenv \
-  && tar czf 3.10.tar.gz versions/$PY_3_10
+  && cd /opt/hostedtoolcache/Python \
+  && tar czf $PY_3_10.tar.gz $PY_3_10
 
 FROM python-core
-RUN pyenv install $PY_3_11 \
+# The pre-compiled Python expects to be installed to this dir
+RUN mkdir -p /opt/hostedtoolcache/Python/$PY_3_11/x64/ \
+  && cd /opt/hostedtoolcache/Python/$PY_3_11/x64/ \
+  # TODO: Add support for arm64 on Ubuntu whenever actions/python-versions adds support for it. Currently not available.
```

Comment:
I expect support for this to be added at some point, eg this upstream PR:
* https://github.com/actions/python-versions/pull/230

At that point, it'll be easy enough to make the `x64` vs `arm64` value dynamic here using `TARGETARCH`...

Reply:
I can't build this locally on the M1 chip. Did you consider copying out the built Python from the official Docker images like how we're installing Ruby? That way we get the architectures for free without extra scripting, should simplify this part.

Reply:
I actually _did_ consider it, but for some reason thought that we were trying to move away from copying from official docker images... but a little digging and I see it's the opposite and the Ruby flip was to move away from `apt-get` towards docker images.

It's a great idea, let me try it.

Reply:
Turned out to be pretty straightforward:
* https://github.com/dependabot/dependabot-core/pull/7934

And it's also a bit cleaner as it allowed copying the files directly into `/usr/local/.pyenv/versions`, no need for any confusing symlinks.